### PR TITLE
Create Spec refactor_gluster_integration_get_state_dump

### PR DESCRIPTION
**Pull request by r0h4n**

### CODE REVIEW

**Comment by r0h4n** [comment]
@brainfunked @nthomas-redhat @anivargi Please review this spec

**Comment by r0h4n** [comment]
@brainfunked @nthomas-redhat Please review

@shtripat Please comment, since you will be taking this up

File: specs/refactor_gluster_integration_get_state_dump.rst [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+Notes:
+
+* In light of new changes to gluster get-state
+  http://review.gluster.org/#/c/15889
+
+Problem description
+===================
+
+Current Tendrl way of consuming and processing is not robust enough to adopt
+with the Tendrl object namespacing, We also need to accomodate lates
```
Comment (shtripat) [inline specs/refactor_gluster_integration_get_state_dump.rst]: %s/lates/latest/g

**Review by shtripat — COMMENTED** [review]
Looks good. One small spelling mistake.
Should I create a github issues for the same and attach here?

File: specs/refactor_gluster_integration_get_state_dump.rst [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+=========
+
+This addresses the use case of obtaining refereshed gluster cluster state
+
+Proposed change
+===============
+
+* General re-factoring of code handling gluster get-state
+
+* Accomodate latest changes to output of gluster get-state w.r.t volume options
```
Comment (cybernth) [inline specs/refactor_gluster_integration_get_state_dump.rst]: Probably more details are required here.  Current format(provided by glusterd) and the expected change

File: specs/refactor_gluster_integration_get_state_dump.rst [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+------------------
+
+None
+
+Other deployer impact
+---------------------
+
+None
+
+Developer impact
```
Comment (cybernth) [inline specs/refactor_gluster_integration_get_state_dump.rst]: What needs to go in here?

File: specs/refactor_gluster_integration_get_state_dump.rst [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+None
+
+Data model impact
+-----------------
+
+* All Persistor models will be affected , especially Volume
+
+
+REST API impact
```
Comment (cybernth) [inline specs/refactor_gluster_integration_get_state_dump.rst]: I think its better to have a section called "IMPACTED MODULES" and list the details over there like REST API/NODE AGENT/COMMON/MONITORING/ALERTING etc if any(only if its affected). make sense?

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/refactor_gluster_integration_get_state_dump.rst (hunk 4 above)
Comment (r0h4n) [reply specs/refactor_gluster_integration_get_state_dump.rst]: Makes sense, please send a PR to change specs/sample_template.rst

File: specs/refactor_gluster_integration_get_state_dump.rst (hunk 3 above)
Comment (r0h4n) [reply specs/refactor_gluster_integration_get_state_dump.rst]: If this spec would require a change in module/tooling/misc, and if other developer's current work/reviews are impacted by this, that goes here

File: specs/refactor_gluster_integration_get_state_dump.rst (hunk 2 above)
Comment (r0h4n) [reply specs/refactor_gluster_integration_get_state_dump.rst]: Sure, its already there in the top, ill add here

File: specs/refactor_gluster_integration_get_state_dump.rst (hunk 1 above)
Comment (r0h4n) [reply specs/refactor_gluster_integration_get_state_dump.rst]: ack

**Review by shtripat — APPROVED** [review]

File: specs/refactor_gluster_integration_get_state_dump.rst (hunk 4 above)
Comment (cybernth) [reply specs/refactor_gluster_integration_get_state_dump.rst]: Sure, I will take care

File: specs/refactor_gluster_integration_get_state_dump.rst (hunk 3 above)
Comment (cybernth) [reply specs/refactor_gluster_integration_get_state_dump.rst]: alright

**Review by cybernth — DISMISSED** [review]

**Comment by r0h4n** [comment]
@brainfunked This needs to be reviewed

**Comment by shtripat** [comment]
The current output from gluster get-state command is of below format 

```
[Volumes]
Volume1.name: test-vol
Volume1.id: 7942d008-e300-4fd9-8af0-5a118afd8d3d
Volume1.type: Distribute
Volume1.transport_type: tcp
Volume1.status: Started
Volume1.brickcount: 1
Volume1.Brick1.path: 172.17.0.2:/tmp/b1
Volume1.Brick1.hostname: 172.17.0.2
Volume1.Brick1.port: 49153
Volume1.Brick1.rdma_port: 0
Volume1.Brick1.status: Started
Volume1.Brick1.signedin: True
Volume1.snap_count: 0
Volume1.stripe_count: 1
Volume1.replica_count: 1
Volume1.subvol_count: 1
Volume1.arbiter_count: 0
Volume1.disperse_count: 0
Volume1.redundancy_count: 0
Volume1.quorum_status: not_applicable
Volume1.snapd_svc.online_status: Offline
Volume1.snapd_svc.inited: True
Volume1.rebalance.id: 00000000-0000-0000-0000-000000000000
Volume1.rebalance.status: not_started
Volume1.rebalance.failures: 0
Volume1.rebalance.skipped: 0
Volume1.rebalance.lookedup: 0
Volume1.rebalance.files: 0
Volume1.rebalance.data: 0Bytes
[Volume1.options]
features.barrier: on
transport.address-family: inet
performance.readdir-ahead: on
nfs.disable: on

Volume2.name: test-vol1
Volume2.id: 35854708-bb72-45a5-bdbd-77c51e5ebfb9
Volume2.type: Distribute
Volume2.transport_type: tcp
Volume2.status: Started
Volume2.brickcount: 1
Volume2.Brick1.path: 172.17.0.2:/tmp/b2
Volume2.Brick1.hostname: 172.17.0.2
Volume2.Brick1.port: 49152
Volume2.Brick1.rdma_port: 0
Volume2.Brick1.status: Started
Volume2.Brick1.signedin: True
Volume2.snap_count: 0
Volume2.stripe_count: 1
Volume2.replica_count: 1
Volume2.subvol_count: 1
Volume2.arbiter_count: 0
Volume2.disperse_count: 0
Volume2.redundancy_count: 0
Volume2.quorum_status: not_applicable
Volume2.snapd_svc.online_status: Offline
Volume2.snapd_svc.inited: True
Volume2.rebalance.id: 00000000-0000-0000-0000-000000000000
Volume2.rebalance.status: not_started
Volume2.rebalance.failures: 0
Volume2.rebalance.skipped: 0
Volume2.rebalance.lookedup: 0
Volume2.rebalance.files: 0
Volume2.rebalance.data: 0Bytes
[Volume2.options]
features.barrier: on
transport.address-family: inet
performance.readdir-ahead: on
nfs.disable: on
```

Parsing logic in tendrl now forms the sections out of this ini as below

**Section: Volumes**
```
Volume1.name: test-vol
Volume1.id: 7942d008-e300-4fd9-8af0-5a118afd8d3d
Volume1.type: Distribute
Volume1.transport_type: tcp
Volume1.status: Started
Volume1.brickcount: 1
Volume1.Brick1.path: 172.17.0.2:/tmp/b1
Volume1.Brick1.hostname: 172.17.0.2
Volume1.Brick1.port: 49153
Volume1.Brick1.rdma_port: 0
Volume1.Brick1.status: Started
Volume1.Brick1.signedin: True
Volume1.snap_count: 0
Volume1.stripe_count: 1
Volume1.replica_count: 1
Volume1.subvol_count: 1
Volume1.arbiter_count: 0
Volume1.disperse_count: 0
Volume1.redundancy_count: 0
Volume1.quorum_status: not_applicable
Volume1.snapd_svc.online_status: Offline
Volume1.snapd_svc.inited: True
Volume1.rebalance.id: 00000000-0000-0000-0000-000000000000
Volume1.rebalance.status: not_started
Volume1.rebalance.failures: 0
Volume1.rebalance.skipped: 0
Volume1.rebalance.lookedup: 0
Volume1.rebalance.files: 0
Volume1.rebalance.data: 0Bytes
```

**Section: Volume1.options**
```
features.barrier: on
transport.address-family: inet
performance.readdir-ahead: on
nfs.disable: on

Volume2.name: test-vol1
Volume2.id: 35854708-bb72-45a5-bdbd-77c51e5ebfb9
Volume2.type: Distribute
Volume2.transport_type: tcp
Volume2.status: Started
Volume2.brickcount: 1
Volume2.Brick1.path: 172.17.0.2:/tmp/b2
Volume2.Brick1.hostname: 172.17.0.2
Volume2.Brick1.port: 49152
Volume2.Brick1.rdma_port: 0
Volume2.Brick1.status: Started
Volume2.Brick1.signedin: True
Volume2.snap_count: 0
Volume2.stripe_count: 1
Volume2.replica_count: 1
Volume2.subvol_count: 1
Volume2.arbiter_count: 0
Volume2.disperse_count: 0
Volume2.redundancy_count: 0
Volume2.quorum_status: not_applicable
Volume2.snapd_svc.online_status: Offline
Volume2.snapd_svc.inited: True
Volume2.rebalance.id: 00000000-0000-0000-0000-000000000000
Volume2.rebalance.status: not_started
Volume2.rebalance.failures: 0
Volume2.rebalance.skipped: 0
Volume2.rebalance.lookedup: 0
Volume2.rebalance.files: 0
Volume2.rebalance.data: 0Bytes
```

**Section: Volume2.options**
```
features.barrier: on
transport.address-family: inet
performance.readdir-ahead: on
nfs.disable: on

<Next volume details>
```

I dont think there is an option to have sub-sections and so the volume details are actually looking screwed up here.

Also with other option where volume option details are listed with volume details only as flattened naming e.g. "Volume1.options.nfs.enabled: on", there is no change required in parsing logic and we can easily parse the options details and push to etcd as a dict. 

For this flattened option, we would require support from gluster to change the output of get-state command.

**Comment by r0h4n** [comment]
@shtripat, Since you own this one,  I am closing this PR, please send a new spec (based on this spec) with whatever changes you feel are necessary to get this spec done.

**Pull request closed without merging**